File: downloads/code/LeetCode/Java/Valid-Palindrome.java
public class Solution {
    public boolean isPalindrome(String s) {
        if (s == null)
            return true;

        int l = 0;
        int r = s.length()-1;
        while (l < r){
            while ((l < r) && !(Character.isLetterOrDigit(s.charAt(l))))
                l += 1;
            while ((l < r) && !(Character.isLetterOrDigit(s.charAt(r))))
                r -= 1;
            if (Character.toLowerCase(s.charAt(l)) != Character.toLowerCase(s.charAt(r)))
                return false;
            l += 1;
            r -= 1;
        }     
        return true;
    }
}

File: downloads/code/LeetCode/Python/Valid-Palindrome.py
class Solution(object):
    def isPalindrome(self, s):
        """
        :type s: str
        :rtype: bool
        """
        if not s:
            return True
        s = s.lower()
        l = 0
        r = len(s)-1
        alpha = "0123456789abcdefghijklmnopqrstuvwxyz"
        while l<r:
            while l < r and s[l] not in alpha:
                l += 1
            if l >= r:
                return True

            while l<r and s[r] not in alpha:
                r -= 1

            if l == r:
                return True

            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True
